lib/charms/tempo_coordinator_k8s/v0/tempo_api.py

```diff
@@ -0,0 +1,219 @@
+        """
+        relations = self.relations
+        if len(relations) == 0:
+            return None
+        if len(relations) > 1:
+            # TODO: Different exception type?
+            raise ValueError("Cannot get_info when more than one application is related.")
+
+        raw_data_dict = relations[0].data.get(relations[0].app)
+        if not raw_data_dict:
+            return None
+
+        # Static analysis errors saying the keys may not be strings.  Protect against this by converting them.
+        raw_data_dict = {str(k): v for k, v in raw_data_dict.items()}
+
+        return TempoApiAppData.model_validate_json(json.dumps(raw_data_dict))  # type: ignore
+
+
+class TempoApiProvider:
+    """The provider side of the tempo-api relation."""
+
+    def __init__(
+        self,
+        relations: RelationMapping,
+        app: Application,
+        direct_url: AnyHttpUrl,
+        ingress_url: Optional[AnyHttpUrl] = None,
+        relation_name: str = DEFAULT_RELATION_NAME,
+    ):
+        """Initialize the TempoApiProvider object.
+
+        This object is for serializing and sending data to a relation that uses the tempo-api interface - it does
+        not automatically observe any events for that relation.  It is up to the charm using this to call publish when
+        it is appropriate to do so, typically on at least the charm's leader_elected event and this relation's
+        relation_joined event.
+
+        Args:
+            relations: The RelationMapping of a charm (typically `self.model.relations` from within a charm object).
+            app: This application.
+            direct_url: The cluster-internal URL at which this application can be reached.  Typically, this is a
+                        Kubernetes FQDN like name.namespace.svc.cluster.local for connecting to the prometheus api
+                        from inside the cluster, with scheme.
+            ingress_url: The non-internal URL at which this application can be reached.  Typically, this is an ingress
+                         URL.
+            relation_name: The name of the relation.
+        """
+        self._charm_relation_mapping = relations
+        self._data = TempoApiAppData(ingress_url=ingress_url, direct_url=direct_url)
+        self._app = app
+        self._relation_name = relation_name
+
+    @property
+    def relations(self):
+        """Return the applications related to us under the monitored relation."""
+        return self._charm_relation_mapping.get(self._relation_name, ())
+
+    def publish(self):
+        """Post tempo-api to all related applications.
+
+        This method writes to the relation's app data bag, and thus should never be called by a unit that is not the
+        leader otherwise ops will raise an exception.
+        """
+        info_relations = self.relations
+        for relation in info_relations:
+            databag = relation.data[self._app]
```

Comment:
also here, there's no such thing as 'too failsafe' when working with relations. 
I'd consider adding this from traefik's ingress; it hasn't let me down yet https://github.com/canonical/traefik-k8s-operator/blob/main/lib/charms/traefik_k8s/v2/ingress.py#L541